File: LLD/src/main/java/org/example/p1_oops/c6_abstract/base/Shape.java
package org.example.p1_oops.c6_abstract.base;

public abstract class Shape {
    protected String name;

    public Shape(String name) {
        this.name = name;
    }

    // Abstract method - MUST be implemented by child classes
    public abstract double getArea();

    // Concrete method - CAN be used directly by child classes
    public void display() {
        System.out.println("Shape: " + name + ", Area: " + getArea());
    }
}

File: LLD NodeJs/p1_oops/c6_abstract/base/Shape.js
/**
 * Abstract Shape class in JavaScript
 * 
 * JavaScript doesn't have the 'abstract' keyword.
 * We simulate abstract classes by:
 * 1. Throwing errors in methods that must be overridden
 * 2. Checking the constructor to prevent direct instantiation (optional)
 */
class Shape {
    _name;

    constructor(name) {
        // Prevent direct instantiation of abstract class
        if (new.target === Shape) {
            throw new Error("Cannot instantiate abstract class Shape directly");
        }
        this._name = name;
    }

    // Abstract method - MUST be implemented by child classes
    getArea() {
        throw new Error("Method 'getArea()' must be implemented by subclass");
    }

    // Concrete method - CAN be used directly by child classes
    display() {
        console.log("Shape: " + this._name + ", Area: " + this.getArea());
    }
}

module.exports = Shape;
